File: src/test/java/com/itsz/portfolio/repository/SecurityRepositoryTest.java
package com.itsz.portfolio.repository;

import com.itsz.portfolio.entity.Security;
import com.itsz.portfolio.entity.SecurityType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@DataJpaTest
public class SecurityRepositoryTest {

    @Autowired
    private SecurityRepository securityRepository;

    @BeforeEach
    public void setUp() {
        securityRepository.deleteAll();
    }

    @Test
    public void testFindByIdentifierStartsWith() {
        Security security1 = new Security("AAPL", SecurityType.STOCK,  null, null, null);
        Security security2 = new Security("AAPL-OCT-2020-110-C",  SecurityType.CALL, new BigDecimal("110.0"), null, LocalDate.now().plusDays(30));
        Security security3 = new Security("GOOGL",  SecurityType.STOCK, null, null, null);
        securityRepository.saveAll(List.of(security1, security2, security3));

        List<Security> result = securityRepository.findByIdentifierStartsWith("AAPL");

        assertEquals(2, result.size());
        assertTrue(result.stream().anyMatch(s -> s.getIdentifier().equals("AAPL")));
        assertTrue(result.stream().anyMatch(s -> s.getIdentifier().equals("AAPL-OCT-2020-110-C")));
    }

    @Test
    public void testFindByIdentifierStartsWith_NoMatches() {
        Security security1 = new Security("AAPL", SecurityType.STOCK, null, null, null);
        Security security2 = new Security("AAPL-OCT-2020-110-C",  SecurityType.CALL, new BigDecimal("110.0"), null, LocalDate.now().plusDays(30));
        securityRepository.saveAll(List.of(security1, security2));

        List<Security> result = securityRepository.findByIdentifierStartsWith("GOOGL");

        assertTrue(result.isEmpty());
    }

    @Test
    public void testFindByIdentifierStartsWith_EmptyDatabase() {
        List<Security> result = securityRepository.findByIdentifierStartsWith("AAPL");
        assertTrue(result.isEmpty());
    }
}

File: src/test/kotlin/com/itsz/portfolio/repository/SecurityRepositoryTest.kt
package com.itsz.portfolio.repository

import com.itsz.portfolio.entity.Security
import com.itsz.portfolio.entity.SecurityType
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import java.math.BigDecimal
import java.time.LocalDate

@DataJpaTest
class SecurityRepositoryTest {

    @Autowired
    private lateinit var securityRepository: SecurityRepository

    @BeforeEach
    fun setUp() {
        securityRepository.deleteAll()
    }

    @Test
    fun `find by identifier starts with returns matching securities`() {
        val security1 = Security("AAPL", SecurityType.STOCK, null, null, null)
        val security2 = Security("AAPL-OCT-2020-110-C", SecurityType.CALL, BigDecimal("110.0"), null, LocalDate.now().plusDays(30))
        val security3 = Security("GOOGL", SecurityType.STOCK, null, null, null)
        securityRepository.saveAll(listOf(security1, security2, security3))

        val result = securityRepository.findByIdentifierStartsWith("AAPL")

        assertEquals(2, result.size)
        assertTrue(result.any { it.identifier == "AAPL" })
        assertTrue(result.any { it.identifier == "AAPL-OCT-2020-110-C" })
    }

    @Test
    fun `find by identifier starts with returns empty when no matches`() {
        val security1 = Security("AAPL", SecurityType.STOCK, null, null, null)
        val security2 = Security("AAPL-OCT-2020-110-C", SecurityType.CALL, BigDecimal("110.0"), null, LocalDate.now().plusDays(30))
        securityRepository.saveAll(listOf(security1, security2))

        val result = securityRepository.findByIdentifierStartsWith("GOOGL")

        assertTrue(result.isEmpty())
    }

    @Test
    fun `find by identifier starts with returns empty when database is empty`() {
        val result = securityRepository.findByIdentifierStartsWith("AAPL")
        assertTrue(result.isEmpty())
    }
}
